Added STORAGE_FORMAT_VERSION and check and updated version check

chupman opened the pull request:
WIP note Currently will only throw an error if an existing 0.2.1 or earlier graph is opened by a v0.3.0 client or server. If a gremlin server is hosting a 0.2.0 graph the client will not throw an error when it connects, which still needs to be addressed.

fixes #232. I differed slightly from the gist linked in issues 232. I'm counting all storage format versions up through 0.2.0 as version 1. JanusGraph versions 0.2.1 and 0.3.0 are storage format 2. I'm hoping to get feedback on the implementation. I'm fully expecting to write up some documentation and figure out how to add test coverage. I'm totally new to Java so please provide brutal feedback to help me improve. 

I'm not really sure how to address titan compatibility right now. Once we officially have a new storage format version it would make sense to leverage any upgrade tooling that is created. 

Usage examples:
Connecting to a existing graph where graph.storage-version is not set.
```
gremlin> graph = JanusGraphFactory.build().set("storage.backend", "cql").open();
13:41:53 WARN  com.datastax.driver.core.NettyUtil  - Found Netty's native epoll transport, but not running on linux-based operating system. Using NIO instead.
Storage format version is incompatible with current client: graph storage version [1] vs. client storage version [2]
Type ':help' or ':h' for help.
Display stack trace? [yN]n
gremlin>
```


Connecting to a fresh Cassandra docker container to show storage-version being set on creation
```
gremlin> graph = JanusGraphFactory.build().set("storage.backend", "cql").open();
13:46:51 WARN  com.datastax.driver.core.NettyUtil  - Found Netty's native epoll transport, but not running on linux-based operating system. Using NIO instead.
==>standardjanusgraph[cql:[127.0.0.1]]
gremlin> mgmt = graph.openManagement()
==>org.janusgraph.graphdb.database.management.ManagementSystem@7209ffb5
gremlin> mgmt.get('graph.storage-version')
==>2
```
Signed-off-by: Chris Hupman <chupman@us.ibm.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there an issue associated with this PR? Is it referenced in the commit message?
- [X] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

chupman commented:
Retargeting at master since no schema changes will be added to 0.2.1(I've been doing too many doc fixes). This PR is in support of the schema updates in #923. 

Separately after talking with @pluradj a utility to update janusgraph-version will need to be created to upgrade an existing graph(probably a bash script). Not sure is it should be added to this PR or created separately.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/configuration/ModifiableConfiguration.java:
```diff
@@ -34,7 +34,8 @@ public ModifiableConfiguration(ConfigNamespace root, WriteConfiguration config,
     public<O> ModifiableConfiguration set(ConfigOption<O> option, O value, String... umbrellaElements) {
         verifyOption(option);
-        Preconditions.checkArgument(!option.isFixed() || !isFrozen(), "Cannot change configuration option: %s", option);
+        Preconditions.checkArgument(option.getName() == "storage-version" ||
```
1. I would use a constant for `"storage-version"` (something like `private final static String STORAGE_VERSION = "storage-version";`). 
2. Are you sure that `option.getName()` is interned? I would use `STORAGE_VERSION.equals(option.getName())` if we don't need much speed in this method. If we do need to increase speed in this method I would change `ConfigElement` constructor line `this.name = name;` to line `this.name = name.intern();` (but it has to be reviewed. I'm not sure.)

pluradj replied:
+1 on `STORAGE_VERSION.equals(option.getName())` for the string comparison, but it's not good to have the upgradeable options hardcoded in this class. It would be better to pass in a list of upgradeable options on a constructor.

chupman commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementSystem.java:
```diff
@@ -175,7 +175,8 @@ public ManagementSystem(StandardJanusGraph graph, KCVSConfiguration config, Log
         @Override
         public void verifyModification(ConfigOption option) {
-            Preconditions.checkArgument(option.getType() != ConfigOption.Type.FIXED, "Cannot change the fixed configuration option: %s", option);
+            Preconditions.checkArgument (graph.getConfiguration().getAllowUpgrade() && option.getName() == "storage-version" || 
```
I was debating creating a separate function to check the option name in case we wanted to add other options, like `janusgraph-version`. I think it might be good to update `janusgraph-version` as well as the storage-version so that old clients that still have the compatible versions check will realize that the graph it's connecting to is not entirely compatible. 


The check for the option name would instead look like
`upgradableElement(option.getName())`
function would probably look like:
```
private boolean upgrableableElement(name) {
    private String[] upgradable = new String[] {"storage-version", "janusgraph-version"};
    return Arrays.asList(upgradeable).contains(name);
}
```

A similar function would have to be added to ModifiableConfiguration.java.

pluradj commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1288,15 +1302,35 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                            String storageVersion = (globalWrite.has(INITIAL_STORAGE_VERSION)) ? globalWrite.get(INITIAL_STORAGE_VERSION) : "1";
+                            throw new JanusGraphException(String.format(INCOMPATIBLE_STORAGE_VERSION_EXCEPTION, storageVersion, JanusGraphConstants.STORAGE_VERSION, storeManager.getName()));
+		        }
+		    } catch(IllegalStateException ise) {
```
I'm not clear if anything in this `try` block throws an `IllegalStateException`. What error scenario are you trying to catch here?

pluradj commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1288,15 +1302,35 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+
+                if (!allowUpgrade) {
+		    try {
+                        if (globalWrite.has(INITIAL_STORAGE_VERSION) && globalWrite.get(INITIAL_STORAGE_VERSION).equals(JanusGraphConstants.STORAGE_VERSION)) {
```
I don't think we're particularly interested in the common case where `INITIAL_STORAGE_VERSION == STORAGE_VERSION`, so no need to log it. Just handle other case.

pluradj commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1288,15 +1302,35 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                globalWrite.set(ALLOW_UPGRADE, allowUpgrade);
+
+                if (!allowUpgrade) {
+		    try {
```
please use consistent indentation here, no tabs.

chupman commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1261,14 +1271,18 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+		
+                //Write storage version
+                Preconditions.checkArgument(!globalWrite.has(INITIAL_STORAGE_VERSION),"Database has already been initialized but not frozen");
+		globalWrite.set(INITIAL_STORAGE_VERSION,JanusGraphConstants.STORAGE_VERSION);
```
Sorry about that. I just fixed my vimrc so it shouldn't happen again. I'll have all all the tabs removed in the next iteration.

chupman commented:
Since I was already checking for tabs, I fixed all the tabs in the files this PR touches. I also addressed most(possibly all?) of the doc fix comments. 
I added a List<String> constant for fixed fields that can be upgraded and moved the checks into separate functions that evaluates based off the new UPGRADEABLE_FIXED constant. 

I gave getCompatibleVersions  a more generic name of getPropertyList so it made sense for populating both functions it's being used for now. 

I'm going to work on automatically updating next, but wanted to commit what I had. I'll circle back on the documentation after the auto-upgrade is done since it will obviously affect the instructions.

chupman commented:
Added auto upgrade and Removed try block. I threw a TODO in about how to handle post upgrade. I'm back and forth between:
1. logging a warning with instructions to remove `graph.allow-upgrade=true`
2. Closing the graph with a message telling users to remove `graph.allow-upgrade=true` from their properties file. 
As I'm typing this I'm starting to lean towards logging a warning. The only unintended consequence is that users could manually edit their storage-version and janusgraph-version with the management API if they leave the setting in place.

chupman commented:
All of the CI failures are timeout related. Can I get someone to rerun the 2 failures? I had different timeout related failures on [my fork](https://travis-ci.org/chupman/janusgraph/builds/389475894), but those 2 did pass.

pluradj commented:
@chupman could you please rebase this with the latest on `master`? thanks.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/JanusGraphConstants.java:
```diff
@@ -60,6 +55,18 @@
+    /**
+     * List of FIXED fields that can be modified when graph.allow-upgrade is set to true 
+     */
+    public static final List<String> UPGRADEABLE_FIXED;
```
As I understand, it should be `Set` and not `List`.

chupman replied:
Thanks for the review. I was trying to avoid adding any additional imports, but I agree a set would perform slightly better. I applied all the changes you suggested and will push them once I do a test upgrade locally.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/configuration/ModifiableConfiguration.java:
```diff
@@ -60,6 +63,15 @@ public void freezeConfiguration() {
     }
 
+    private boolean upgradeFixed(String name) {
+        for (String value : UPGRADEABLE_FIXED) {
```
1. This loop isn't needed because `List` has a method `contains`. 
2. You should use `HashSet` to make complexity of `upgradeFixed` be `O(1)` and not `O(n)`.
3. It would be clearer if the name of checking method would start from `is` or `has`.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1716,6 +1760,17 @@ public TimestampProvider getTimestampProvider() {
         return configuration.get(TIMESTAMP_PROVIDER);
     }
 
+    public boolean getAllowUpgrade(String name) {
```
should start with `is` or `has`

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1716,6 +1760,17 @@ public TimestampProvider getTimestampProvider() {
 
+    public boolean getAllowUpgrade(String name) {
+        if (configuration.get(ALLOW_UPGRADE)) {
+            for (String value : JanusGraphConstants.UPGRADEABLE_FIXED) {
```
It is a duplicate code. I think you should use something like:
```
public boolean isUpgradeAllowed(String name) {
  return configuration.get(ALLOW_UPGRADE) && JanusGraphConstants.UPGRADEABLE_FIXED.contains(name);
}
```

pluradj commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1286,15 +1301,46 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                } else {
+                         if (!globalWrite.has(INITIAL_STORAGE_VERSION) || !globalWrite.get(INITIAL_STORAGE_VERSION).equals(JanusGraphConstants.STORAGE_VERSION)) {
+                            String storageVersion = (globalWrite.has(INITIAL_STORAGE_VERSION)) ? globalWrite.get(INITIAL_STORAGE_VERSION) : "1";
+                            throw new JanusGraphException(String.format(INCOMPATIBLE_STORAGE_VERSION_EXCEPTION, storageVersion, JanusGraphConstants.STORAGE_VERSION, getGraphName()));
```
`getGraphName()` can return `null`. It would be a nicer error message if it used an empty string rather than `null`.

pluradj commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1286,15 +1301,46 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                                 globalWrite.get(INITIAL_STORAGE_VERSION), JanusGraphConstants.STORAGE_VERSION, globalWrite.get(INITIAL_JANUSGRAPH_VERSION), JanusGraphConstants.VERSION, getGraphName());
+                    // If the storage version of the client or server opening the graph is lower than the graph's storage version throw an exception
+                    } else if (Integer.parseInt(globalWrite.get(INITIAL_STORAGE_VERSION)) > Integer.parseInt(JanusGraphConstants.STORAGE_VERSION)) {
+                        throw new JanusGraphException(String.format(BACKLEVEL_STORAGE_VERSION_EXCEPTION, globalWrite.get(INITIAL_STORAGE_VERSION), JanusGraphConstants.STORAGE_VERSION, getGraphName()));
```
`getGraphName()` can return `null`. It would be a nicer error message if it used an empty string rather than `null`.

pluradj commented on docs/upgrade.adoc:
```diff
@@ -5,6 +5,69 @@
+==>0.3.0-SNAPSHOT
+mgmt.close()
+==>null
+----
```
I think you can remove this code block now since updating the Management API is no longer necessary.

pluradj commented on docs/upgrade.adoc:
```diff
@@ -5,6 +5,62 @@
+
+----
+
+*IMPORTANT* You should back-up your data prior to attempting an upgrade!
```
A migration from Titan, however, is unable to detect that the storage format has changed since Titan would not know how to determine this (`titan.version` and `titan.compatible-versions` are essentially hardcoded in the Titan runtime).

Please add text to the [Migrating from Titan](https://github.com/JanusGraph/janusgraph/blob/master/docs/migrating.adoc) documentation to warn them that once they have migrated to JanusGraph 0.3.0 that they should not connect to the graph instance again with Titan.

chupman replied:
I was thinking I could also add the line below to the upgrade to prevent titan connections
```
globalWrite.set(TITAN_COMPATIBLE_VERSIONS, JanusGraphConstants.VERSION);
```

TITAN_COMPATIBLE_VERSIONS sets graph.titan-version which lines up with INITIAL_TITAN_VERSION. Titan has a check on graph.titan-version that should prevent connection.

pluradj commented:
@chupman also please rebase to pick up the updates on `JanusGraphGremlinPlugin.java`

pluradj commented on docs/upgrade.adoc:
```diff
@@ -5,6 +5,54 @@
 JanusGraph release.
 
+=== Upgrading to JanusGraph 0.3.0
+*IMPORTANT* You should back-up your data prior to attempting an upgrade! Also please note that once an upgrade has been completed you will no longer be able to connect to your graph with client versions prior to 0.3.0.
```
Please change this `*IMPORTANT*` to `IMPORTANT:` so that it creates a call out box.

chupman replied:
It has been changed to ```IMPORTANT:```

porunov commented:
@chupman Seems now this PR has conflicts. Could you please resolve them?

sjudeng commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1286,15 +1301,54 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                else
+                    allowUpgrade = ALLOW_UPGRADE.getDefaultValue();
+
+                if (allowUpgrade) {
```
Can this block be tested to improve coverage (see [here](https://codecov.io/gh/JanusGraph/janusgraph/src/storage-version/janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java#L1319...1341))?

chupman replied:
I've been working on improving the test coverage in a different branch. GraphDatabaseConfiguration needs to be refactored to make it possible. Would it be ok if I opened a separate issue to track the refactor and test coverage improvement?

sjudeng replied:
Sure thanks for taking a look. There's probably already a separate issue for improving test coverage.

chupman commented:
Thanks @sjudeng. I appreciate the review cycles.

pluradj commented:
Thank you @chupman @porunov @sjudeng
